pkg/controller/externalnode/controller_test.go

```diff
@@ -0,0 +1,652 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package externalnode
+
+import (
+	"context"
+	"crypto/sha1" // #nosec G505: not used for security purposes
+	"encoding/hex"
+	"io"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	"antrea.io/antrea/pkg/client/clientset/versioned"
+	fakeclientset "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions"
+)
+
+var (
+	informerFactory crdinformers.SharedInformerFactory
+)
+
+func TestAddExternalNode(t *testing.T) {
+	for _, tc := range []struct {
+		externalNode     *v1alpha1.ExternalNode
+		expectedEntities []v1alpha2.ExternalEntity
+	}{
```

Comment:
Better to add a name field and use it as the subtest name, which could self-explain the test case and be useful when figuring out which subtest fails, otherwise the test failure would just say which line fails but not which case.
```
t.Run(tc.name, func(t *testing.T) {
    ...
}
```